hooks/environment

```diff
@@ -1,6 +1,19 @@
 #!/usr/bin/env bash
 set -eu -o pipefail
 
+# Some callers slightly abuse this script by including it via Bash's `source` command, instead of calling it as an executable. This means this code is sometimes executed with extra Bash context (namely options and error traps) that change how the code is interpreted. This interference is not always desirable, but unfortunately we must tolerate usage via `source`.
+#
+# First we detect whether the `-E` option is set. This matters because when `-E` is combined with an error trap (as in the agent environment hook), it breaks this hook's retry functionality.
+#
+# The special $- variable is a list of all the options currently set. We find out if it contains 'E', and store the result in IS_BIG_E_OPTION_SET.
+#
+# We'll use IS_BIG_E_OPTION_SET later on, to decide if we need to remove and reinstate it around retries.
+#
+IS_BIG_E_OPTION_SET=no
+if [[ "$-" =~ "E" ]]; then
+  IS_BIG_E_OPTION_SET=yes
+fi
```

Comment:
This is very cool. I would encourage you to share this (and all the other twists and turns this took) in Eng Hangout. I think there's a good story to tell about why the double sourcing occurs, and this knowledge certainly needs to be spread.

Sorry to bikeshed even more on this, but I think calling the variable `WAS_BIG_E_OPTION_SET` conveys its purpose a bit more effectively
```suggestion
WAS_BIG_E_OPTION_SET=no
if [[ "$-" =~ "E" ]]; then
  WAS_BIG_E_OPTION_SET=yes
fi
```